[Blueprints] Login step – handle passwordless autologin via a PHP mu-plugin

bgrgicak opened the pull request:
Fixes #1706

## Motivation for the change, related issues

Handles auto login, e.g. in the `login` Blueprint step, using a platform-provided mu-plugin. This enables using the `login` step across all Playground runtimes without any of the issues with the previous implementation.

Before this PR, the `login` worked by making HTTP requests to the `/wp-login.php` endpoint. That approach had significant downsides:

* It only worked in web browsers
* It didn't support custom login mechanisms
* It required storing plaintext passwords in the Blueprint files
* It broke after a content import if the admin password changed
* It broke if WordPress displayed the [Site Admin Email Verification Screen](https://make.wordpress.org/core/2019/10/17/wordpress-5-3-admin-email-verification-screen/).

## Implementation details

To fix it and make the login step more robust login now uses a login PHP file that logs in the user with their credentials. 
The new script only requires a username, the password is now deprecated.

There are two ways to trigger the auto-login:

### The `PLAYGROUND_AUTO_LOGIN_AS_USER` constant

Used by the `login` Blueprint step does.

When the `PLAYGROUND_AUTO_LOGIN_AS_USER` constant is defined, this mu-plugin
will automatically log the user in on their first visit. The username is
the value of the constant.

On subsequent visits, the `playground_auto_login_happened` cookie will be
detected and the user will not be logged in. This means the "logout" feature
will work as expected.

### The `playground_force_auto_login_as_user` GET parameter

Used by the "login" button in various Playground runtimes.

When the `playground_force_auto_login_as_user` GET parameter is present,
this mu-plugin will automatically log in any logged out visitor. This will
happen every time they visit, not just on their first visit.
 
## Testing Instructions (or ideally a Blueprint)

- [Open this blueprint and confirm that login works](http://127.0.0.1:5400/website-server/?import-site=http%3A%2F%2F127.0.0.1%3A5400%2Fwebsite-server%2Ftest-fixtures%2Fsite-with-unpatched-gutenberg.zip&login=yes&url=%2Fwp-admin%2Fpost-new.php)
- Open Playground, log out, then navigate to `/?playground_force_auto_login_as_user=admin&` and confirm you were logged in again.

cc @brandonpayton – the `/?playground_force_auto_login_as_user=admin` triggers a redirect to add the trailing slash, as in `/?playground_force_auto_login_as_user=admin/`. Let's track that separately.

akirk commented on packages/playground/blueprints/src/lib/steps/login.ts:
```diff
@@ -37,22 +38,34 @@ export const login: StepHandler<LoginStep> = async (
+			return;
+		}
+
+		$user = wp_signon( array(
+			'user_login'    => ${phpVar(username)},
+			'user_password' => ${phpVar(password)},
+			'remember'      => true
+		) );
```
If we do this on the PHP side we could remove the need to use a password and use `wp_set_auth_cookie()` instead of `wp_signon()`. It's not a big deal but with this we are storing the password in clear text.

bgrgicak replied:
I agree, there is a working script in the PR description that would remove the need for passwords.

I'm happy to do it, but I wanted to get feedback first.

@adamziel what do you think about switching to passwordless logins?

akirk replied:
Oh sorry, I had missed that!

adamziel replied:
This came up in another issue, see:

https://github.com/WordPress/wordpress-playground/pull/1753#discussion_r1755602424

Tl;dr let's avoid writing to document root and make it a flexible mu-plugin for all runtimes. That would effectively be a passwordless login

bgrgicak replied:
> Tl;dr let's avoid writing to document root and make it a flexible mu-plugin for all runtimes. That would effectively be a passwordless login

I'm having problems with [making this approach work](https://github.com/WordPress/wordpress-playground/pull/1753#discussion_r1755602424). I assume it's because it's triggered by page loads.

I ended up using my implementation with a HTTP request but it uses auto_prepend_file to add the file instead of writing it inside the step.

adamziel replied:
> I'm having problems with making this approach work. I assume it's because it's triggered by page loads.

What are these problem?

bgrgicak commented on packages/playground/blueprints/src/lib/steps/login.ts:
```diff
@@ -21,44 +19,28 @@ export type LoginStep = {
-			rememberme: 'forever',
-		},
+	// Make a request to the wp-admin to ensure the user is logged after the blueprint runs.
+	await playground.request({
```
@adamziel I couldn't find a good way to access the request handler.

[The example we have doesn't work](https://github.com/WordPress/wordpress-playground/blob/27dc0401e540c0a84f9fd031254714748815f79a/packages/php-wasm/universal/src/lib/php-request-handler.ts#L126-L153).

How can we make this easy?

adamziel replied:
@bgrgicak I'd like to stop using the `request()` method and remove it from the Playground interface accessible in Step handlers. It only makes sense in the browser and isn't portable.

adamziel replied:
In terms of mental models, creating a site with a Blueprint is distinct from accessing the site later on. We'll use Blueprints to create zipped site exports – cookies are not practical in that model. Therefore, let's avoid generating any auth cookies in the step handler. Let's only install the autoloading plugin without actually logging the user in. They'll be logged in the first time they access the Playground site.

brandonpayton replied:
One thing I wonder about here is whether this would break using Playground to test logged-out behavior.

For example, if people wanted to test logged-in and logged-out views of the front end with Playground, how would this work?

bgrgicak replied:
> One thing I wonder about here is whether this would break using Playground to test logged-out behavior.

It will work in the same way it does today. It you use `?login=no` you won't be logged in.

bgrgicak commented:
I'm still having some issues with tests not passing, but when I test the feature in a browser it works.
It's probably related to the node version.

adamziel commented:
@bgrgicak Do you mean this error?

```
AssertionError: expected '' to include 'Dashboard'
```

any chance the test needs to account for a 302 redirect?

adamziel commented:
Also, regarding this failure:

```
"text": "#!/usr/bin/env php
WP-CLI only works correctly from the command line, using the 'cli' PHP SAPI.
You're currently executing the WP-CLI binary via the 'wasm' PHP SAPI.
In case you were trying to run this file with a web browser, know that this cannot technically work.
```

Try passing the `sapiName: 'cli',` argument to `bootWordPress` in the `enable-multisite.spec.ts` test – it's what the WP-CLI step test does.

brandonpayton commented:
> Try passing the `sapiName: 'cli',` argument to `bootWordPress` in the `enable-multisite.spec.ts` test – it's what the WP-CLI step test does.

For the multisite failures, I tried this, and we get another failure. But it seems like progress.

```
FAIL  src/lib/steps/enable-multisite.spec.ts > Blueprint step enableMultisite > should enable a multisite on a scopeless URL
Error: PHP.run() failed with exit code 1 and the following output: Error: This does not seem to be a WordPress installation.
The used path is: /wordpress/    
Pass --path=`path/to/wordpress` or run `wp core download`.
```

Then I tried adding a `--path=/wordpress` arg for WP-CLI, and it still complained about the same thing.

brandonpayton commented:
> @bgrgicak Do you mean this error?
> 
> ```
> AssertionError: expected '' to include 'Dashboard'
> ```
> 
> any chance the test needs to account for a 302 redirect?

@bgrgicak I adjusted the login tests to account for the redirects to and from wp-login.php, but the redirect from wp-login.php is not occurring. Is the test missing something, or is something not working?

I also added an empty test as a test TODO for logged out behavior.

As an unrelated item, I pushed https://github.com/WordPress/wordpress-playground/pull/1789/commits/91ef09640b8702aea2d5f6ba516daae31b0e5039 which stops using a default password for the login step in the main Playground web app and the Blueprint builder.

Please feel free to revert either of these commits if they are not helpful. I was hoping to resolve the remaining test failures before you started your day but ran into further issues.

bgrgicak commented:
@brandonpayton I took a look at the HTML the second request returns the [Site Admin Email Verification Screen](https://make.wordpress.org/core/2019/10/17/wordpress-5-3-admin-email-verification-screen/).

I think that we are trying to force unit tests to behave like E2E tests here and that we should just rewrite these tests to Playwright. 

The login page redirect messes up these tests, but I couldn't find a way around using it.

bgrgicak commented:
This should work now! I changed the failing unit tests to check for constants instead of making HTTP requests.

bgrgicak commented on packages/playground/blueprints/public/blueprint-schema.json:
```diff
@@ -698,6 +698,29 @@
+							},
+							"additionalProperties": false
+						},
+						"step": {
```
I added this to have a [step inside compile](https://github.com/WordPress/wordpress-playground/pull/1789/files#diff-111618e10ba828ef4596cfba7f3a377a0c7cefc4b2432d96a13a9bd77f462987) which we can insert into the blueprint.

I could simply call the function from there, but it felt wrong because we don't execute any steps during compile.

bgrgicak commented on packages/playground/blueprints/src/lib/steps/enable-multisite.spec.ts:
```diff
@@ -25,26 +24,32 @@ describe('Blueprint step enableMultisite', () => {
-			url: '/wp-admin/network/',
+		const result = await php.run({
+			code: `
+				<?php
+				echo WP_ALLOW_MULTISITE ? 'true' : 'false';
+			`,
 		});
-		expect(response.text).toContain('My Sites');
-	}, 30_000);
+		expect(result.text).toContain('true');
```
Relying on HTTP requests for unit tests didn't work anymore because of the `wp-login` redirect, so I updated the test to look for this constant instead. 

E2E tests will still do the full check.

adamziel replied:
Your comment would make a great inline comment. This discussion will be lost in backscroll tomorrow, but a comment would still be there two years from now.

Also, it would be useful to change the name of that test to something like "should set the WP_ALLOW_MULTISITE to true" since the old name no longer reflects what it does, e.g. the constant doesn't tell us whether we're actually running a working multisite.

adamziel replied:
I noticed the test name remains unchanged

bgrgicak replied:
Done [22d7ae8](https://github.com/WordPress/wordpress-playground/pull/1789/commits/22d7ae8d768f4cd10bccda9a911fbb32f995f8f7)

bgrgicak commented on packages/playground/blueprints/src/lib/steps/enable-multisite.ts:
```diff
@@ -34,6 +30,8 @@ export interface EnableMultisiteStep {
 export const enableMultisite: StepHandler<EnableMultisiteStep> = async (
 	playground
 ) => {
+	await installWpCli(playground, {});
```
This will allow JS API users to execute commands even if they didn't include the `extraLibrary`.

bgrgicak commented on packages/playground/blueprints/src/lib/compile.ts:
```diff
@@ -192,21 +202,8 @@ export function compileBlueprint(
-				 *        like the JavaScript parser or Composer are removed.
-				 */
-				url: 'https://playground.wordpress.net/wp-cli.phar',
-			},
-			path: '/tmp/wp-cli.phar',
+			data: wpCliResource,
+			path: wpCliPath,
```
I attempted to reuse data here, but it doesn't feel great.

adamziel replied:
Yeah let's not fetch that plugin in the JS API handler at all. I realize that's inconsistent, let's solve for that later on as we migrate to the PHP library.

bgrgicak commented on packages/playground/blueprints/src/lib/steps/login.spec.ts:
```diff
@@ -25,10 +25,29 @@ describe('Blueprint step installPlugin', () => {
+		expect(result.text).toContain('admin');
+	});
+
+	it('should redirect user to wp-login.php with correct redirect_to', async () => {
```
@brandonpayton I kept only the first part of your test. This combined with a E2E test should be good enough for now.

bgrgicak commented on packages/playground/blueprints/src/lib/steps/login.ts:
```diff
@@ -21,44 +19,23 @@ export type LoginStep = {
- * Under the hood, this function submits the [`wp-login.php`](https://developer.wordpress.org/reference/files/wp-login.php/) [form](https://developer.wordpress.org/reference/functions/wp_login_form/)
- * just like a user would.
+ * Under the hood, this function sets the `PLAYGROUND_AUTO_LOGIN` constant.
+ * The `0-playground.php` mu-plugin uses that constant to log in the user on first load.
```
```suggestion
 * The `auto_login.php` mu-plugin uses that constant to log in the user on the first load.
```

bgrgicak commented on packages/playground/blueprints/src/lib/steps/wp-cli.ts:
```diff
@@ -1,6 +1,34 @@
+export const installWpCli = async (playground: UniversalPHP) => {
+	await writeFile(playground, {
+		data: new File(
+			[await fetch(wpCliResource.url).then((r) => r.arrayBuffer())],
```
Is there a better way to do this or is `fetch` a good fit here?

bgrgicak commented on packages/playground/website/cypress/e2e/query-api.cy.ts:
```diff
@@ -179,7 +179,7 @@ describe('Query API', () => {
 
 	describe('option `multisite`', () => {
-		it('should enable a multisite', () => {
+		it.only('should enable a multisite', () => {
```
```suggestion
		it('should enable a multisite', () => {
```

bgrgicak commented on packages/playground/remote/src/lib/playground-mu-plugin/0-playground.php:
```diff
@@ -150,5 +152,3 @@ function networking_disabled() {
 		return [ 'Dummy' ];
 	});
 }
```
```suggestion
}

?>
```

adamziel commented on packages/playground/blueprints/src/lib/steps/enable-multisite.ts:
```diff
@@ -34,6 +29,8 @@ export interface EnableMultisiteStep {
 export const enableMultisite: StepHandler<EnableMultisiteStep> = async (
 	playground
 ) => {
+	await installWpCli(playground);
```
This will always fetch wp-cli, even when it's already available.

bgrgicak replied:
I forgot to add that part back after making changes. It's fixed now.

adamziel commented on packages/playground/blueprints/src/lib/steps/wp-cli.ts:
```diff
@@ -31,16 +59,7 @@ export const wpCLI: StepHandler<WPCLIStep, Promise<PHPResponse>> = async (
-			Read more about it in the documentation.
-			https://wordpress.github.io/wordpress-playground/blueprints/data-format#extra-libraries`);
-	}
+	await installWpCli(playground);
```
This will always fetch wp-cli, even if it's already available

adamziel replied:
Also, why fetch it at all here instead of leaving that up to the API consumer? This opens the API up for more scenarios we'll have to maintain and migrate to the PHP library. I understand this is not a convenient constraint to follow, but introducing `installWpCli` opens up API a bit more than I'm comfortable with, e.g. now we have multiple places in the codebase that can perform the installation, the `.phar` file won't be downloaded upfront concurrently with other resources, etc.

adamziel commented on packages/playground/blueprints/src/lib/steps/wp-cli.ts:
```diff
@@ -85,7 +106,7 @@ export const wpCLI: StepHandler<WPCLIStep, Promise<PHPResponse>> = async (
 
 	const result = await playground.run({
-		scriptPath: '/wordpress/run-cli.php',
+		scriptPath: `${documentRoot}/run-cli.php`,
```
Nit: let's always use `joinPaths()`

bgrgicak replied:
Done

adamziel commented on packages/playground/remote/src/lib/playground-mu-plugin/playground-includes/auto_login.php:
```diff
@@ -0,0 +1,71 @@
+<?php
+
```
Let's add a file-level comment with thorough context on how and when this code runs

adamziel commented on packages/playground/remote/src/lib/playground-mu-plugin/playground-includes/auto_login.php:
```diff
@@ -0,0 +1,71 @@
+add_action('init', function() {
+	if (strpos($_SERVER['REQUEST_URI'], 'wp-login.php') === false) {
+		return;
+    }
```
nit: tabs vs spaces, we'll need to set up some PHP formatting tool later on (not a blocker here)

adamziel commented on packages/playground/blueprints/src/lib/steps/login.ts:
```diff
@@ -21,44 +19,23 @@ export type LoginStep = {
 	username?: string;
 	/**
-	 * The password to log in with. Defaults to 'password'.
+	 * @deprecated The password field is deprecated and will be removed in a future version.
```
lovely :heart:

bgrgicak commented:
@adamziel I fixed the login issue introduced after adding query param support and added a test for it.
Thank you for your improvements, we should be ready to merge.

adamziel commented on packages/playground/blueprints/src/lib/steps/enable-multisite.spec.ts:
```diff
@@ -36,13 +36,21 @@ describe('Blueprint step enableMultisite', () => {
 		await enableMultisite(php, {});
 
+		/**
+		 * Relying on HTTP requests for unit tests won't work because of the wp-login redirects.
```
can this comment clarify how the redirects prevent the unit test from working? So that someone with no knowledge about this discussion would immediately understand

adamziel commented on packages/playground/wordpress/src/mu-plugins/auto_login.php:
```diff
@@ -1,59 +1,86 @@
+ *
+ * @return string|false
+ */
+function get_playground_auto_login_as_user() {
```
Let's prefer names that read like something a person would say, e.g. `playground_get_username_for_auto_login` (where `playground_` is a prefix to avoid function name collisions).

adamziel commented on packages/playground/blueprints/src/lib/steps/login.spec.ts:
```diff
@@ -27,7 +28,30 @@ describe('Blueprint step login', () => {
+		expect(adminResponse.text).toContain('Dashboard');
+	});
+
+	it('should log the user in if the auto login parameter is set', async () => {
```
Nit, let's make this name more precise. When this fails, we'll see the test name in red in CLI. Let's be kind to our future selves and communicate this is about the query string parameter, say:

should log the user in if the `playground_force_auto_login_as_user ` query parameter is set

adamziel commented on packages/playground/wordpress/src/mu-plugins/auto_login.php:
```diff
@@ -106,7 +106,7 @@ function playground_auto_login() {
 	 */
-	if (false !== stripos( wp_login_url(), $_SERVER['SCRIPT_NAME'] )) {
+	if (false === stripos( wp_login_url(), $_SERVER['SCRIPT_NAME'] )) {
 		return;
```
What would happen without this check? Also, stripos doesn't seem like a reliable way of testing for a login page. What would a rigorous, 100% reliable check look like?

bgrgicak replied:
This code is a copy of [is_login](https://developer.wordpress.org/reference/functions/is_login/) U just extracted it because I need to add `empty($_SERVER['SCRIPT_NAME'])`.

> What would happen without this check?

`playground_auto_login` would be called more frequently and users would be redirected on every page load.
But this makes me think that it could all be one hook. I will explore this a bit more and at least add more comments.

adamziel commented on packages/playground/wordpress/src/mu-plugins/auto_login.php:
```diff
@@ -106,7 +106,7 @@ function playground_auto_login() {
 	 * Check if the request is for the login page.
 	 */
-	if (false === stripos( wp_login_url(), $_SERVER['SCRIPT_NAME'] )) {
+	if (empty($_SERVER['SCRIPT_NAME']) || false === stripos( wp_login_url(), $_SERVER['SCRIPT_NAME'] )) {
```
Why would SCRIPT_NAME be empty?

bgrgicak replied:
I tried debugging it and it looks like this code runs even if we don't make an HTTP request.
In these cases `$_SERVER['SCRIPT_NAME']` is empty.

adamziel replied:
Why does it run even if we don't make an HTTP request? It seems like a bug. Silencing this warning only hides the problem. What's the root cause and what's best way of addressing it?

bgrgicak replied:
I will take a look at why this is happening.

bgrgicak commented on packages/playground/blueprints/src/lib/steps/login.spec.ts:
```diff
@@ -26,9 +27,45 @@ describe('Blueprint step installPlugin', () => {
+		expect(adminResponse.text).toContain('Dashboard');
+	});
+
+	it('should log the user in if the auto login parameter is set', async () => {
```
```suggestion
	it('should log the user in if the playground_force_auto_login_as_user query parameter is set', async () => {
```

bgrgicak commented:
I got all the blueprint unit tests to pass locally, but I still need to figure out what's happening with the rest.

bgrgicak commented:
I'm closing this PR in favour of https://github.com/WordPress/wordpress-playground/pull/1856.

Because of some recent force pushing to trunk and me merging trunk into this branch, it picked up 300 extra commits.
